GitUpKit/Interface/GIGraphView.m

```diff
@@ -860,97 +860,168 @@ static void _DrawLine(GILine* line, CGContextRef context, CGFloat offset, CGFloa
+  
+  CFAttributedStringRef string = (CFAttributedStringRef)multilineAttributedTitle.copy;
+  [multilineAttributedTitle release];
+  
+  // Prepare CoreText string from the rich attributed title
   
-  CFAttributedStringRef string = CFAttributedStringCreate(kCFAllocatorDefault, (CFStringRef)label, (CFDictionaryRef)attributes);
   CTFramesetterRef framesetter = CTFramesetterCreateWithAttributedString(string);
   CGSize size = CTFramesetterSuggestFrameSizeWithConstraints(framesetter, CFRangeMake(0, CFAttributedStringGetLength(string)), NULL, CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX), NULL);
   CGRect textRect = CGRectMake(kTitleOffsetX, kTitleOffsetY, ceil(size.width), ceil(size.height));
   CGPathRef path = CGPathCreateWithRect(textRect, NULL);
   CTFrameRef frame = CTFramesetterCreateFrame(framesetter, CFRangeMake(0, CFAttributedStringGetLength(string)), path, NULL);
-  CFAttributedStringRef character = CFAttributedStringCreate(kCFAllocatorDefault, CFSTR("\u2026"), (CFDictionaryRef)attributes);
-  CTLineRef token = CTLineCreateWithAttributedString(character);
+  CFAttributedStringRef ellipsis = CFAttributedStringCreate(kCFAllocatorDefault, CFSTR("\u2026"), (CFDictionaryRef)multilineTitleAttributes);
+  CTLineRef ellipsisToken = CTLineCreateWithAttributedString(ellipsis);
   
   // Prepare context
   
   CGContextSaveGState(context);
   CGContextTranslateCTM(context, x, y);
-  CGContextRotateCTM(context, 40.0 / 180.0 * M_PI);
-  
-  // Draw label
-  
-#if __DEBUG_BOXES__
-  CGRect labelRect = CGRectInset(CGRectMake(textRect.origin.x, textRect.origin.y, MIN(textRect.size.width, kMaxBranchTitleWidth), textRect.size.height), -4.5, -3.5);
-  CGContextSetRGBFillColor(context, 1.0, 0.0, 0.0, 0.666);
-  CGContextFillRect(context, labelRect);
-#endif
+  CGContextRotateCTM(context, 45.0 / 180.0 * M_PI);
   
   // Draw text
   
-  CGContextSetRGBFillColor(context, 0.0, 0.0, 0.0, 1.0);
+  CGFloat lastLineWidth = 0.0;
   CFArrayRef lines = CTFrameGetLines(frame);
   for (CFIndex i = 0, count = CFArrayGetCount(lines); i < count; ++i) {
     CTLineRef line = CFArrayGetValueAtIndex(lines, i);
     CGPoint origin;
     CTFrameGetLineOrigins(frame, CFRangeMake(i, 1), &origin);
-    CGContextSetTextPosition(context, textRect.origin.x + origin.x, textRect.origin.y + origin.y);
-    if (size.width <= kMaxBranchTitleWidth) {
+    
+    origin.x += textRect.origin.x + origin.y;
+    origin.y += textRect.origin.y;
+    
+    CGFloat ascent, descent, leading;
+    CGFloat lineWidth = CTLineGetTypographicBounds(line, &ascent, &descent, &leading);
+    CGFloat lineHeight = ascent;
+    
+    // Draw separator in case of new line which is guaranteed by the building algorithm
+    
+    CFRange stringRange = CTLineGetStringRange(line);
+    if (stringRange.length == 1) {
+      CGRect underlineRect = CGRectMake(origin.x - 1.0, origin.y - 1.0, lastLineWidth + 5.0, lineHeight - 3.0);
+      CGMutablePathRef separatorPath = CGPathCreateMutable();
```

Comment:
You don't need to build a path here and then stoke it. Just build the path directly in the context and stroke it. It's quite simpler :smile: 
```
CGContextMoveToPoint()
CGContextAddLineToPoint()
...
CGContextStrokePath()
```

You don't need to save/restore the state when touching paths either.